src/common/base/src/base/dma.rs

```diff
@@ -0,0 +1,508 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::alloc::alloc;
+use std::alloc::dealloc;
+use std::alloc::Layout;
+use std::io;
+use std::io::IoSlice;
+use std::io::SeekFrom;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::ops::Range;
+use std::os::fd::BorrowedFd;
+use std::os::unix::io::AsRawFd;
+use std::path::Path;
+
+use rustix::fs::OFlags;
+use tokio::fs::File;
+use tokio::io::AsyncSeekExt;
+
+use crate::runtime::spawn_blocking;
+
+/// An aligned buffer used to perform io on a `DmaFile`.
+#[derive(Debug)]
+pub struct DmaBuffer {
+    cap: usize,
+    len: usize,
+    align: usize,
+    data: *mut u8,
+}
+
+unsafe impl Send for DmaBuffer {}
+
+impl DmaBuffer {
```

Comment:
It would be better to remove **some unsafe code** in `DmaBuffer` and we can refactor it by using `Vec<u8>`.